sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.shuffle
+
+import scala.collection.mutable.ArrayBuffer
+
+// Helper trait that callers can use to add blocks to the iterator
+// as long as they can provide a size
+trait BlockWithSize {
+  /**
+   * Abstract method to return the size in bytes of this block
+   * @return Long - size in bytes
+   */
+  def size: Long
+}
+
+/**
+ * Specifies a start and end range of byes for a block.
+ * @param block - a BlockWithSize instance
+ * @param rangeStart - byte offset for the start of the range
+ * @param rangeEnd - byte offset for the end of the range
+ * @tparam T - the specific type of `BlockWithSize`
+ */
+case class BlockRange[T <: BlockWithSize](
+    block: T, rangeStart: Long, rangeEnd: Long) {
+
```

Comment:
Do we want an assertion to verify that start < end or does it even matter?

Reply:
Can't hurt to add for sure. Note that these are inclusive, if you want a 1 byte window:
```
BlockRange(block, 0, 0)
```

Reply:
The scaladoc should explain that these are both inclusive.  By the way. I find that a bit odd.  Almost everything uses the convention of start=inclusive and end=exclusive, so this will be confusing.  Is there a compelling reason to keep the range end inclusive?  Changing this would also clean up all the +1 / -1 stuff going on which seems like it's begging for an off-by-one error to show up at some point down the road.

Reply:
`end` is now exclusive. `BlockRange(block, 0, 1)` is how one would specify a 1-byte block.